File: javascript/binarySearchTree.js
//functional-shared

var BinarySearchTree = function(value) {
  var newBinarySearchTree = {};
  newBinarySearchTree.value = value;
  newBinarySearchTree.left;
  newBinarySearchTree.right;
  newBinarySearchTree.level = 0;


  newBinarySearchTree.insert = binarySearchTreeMethods.insert;
  newBinarySearchTree.contains = binarySearchTreeMethods.contains;
  newBinarySearchTree.DFSEach = binarySearchTreeMethods.DFSEach;
  newBinarySearchTree.DFSToArray = binarySearchTreeMethods.DFSToArray;
  newBinarySearchTree.BFSEach = binarySearchTreeMethods.BFSEach;
  newBinarySearchTree.BFSToArray = binarySearchTreeMethods.BFSToArray;

  return newBinarySearchTree;
};

var binarySearchTreeMethods = {};

//constant time complexity O(1)
binarySearchTreeMethods.insert = function(input){
    if(input < this.value){
      if(!this.left) {
        this.left = BinarySearchTree(input);
      } else {
        this.left.insert(input);
      }
  } else if(input > this.value){
    if (!this.right) {
      this.right = BinarySearchTree(input);
    } else {
      this.right.insert(input);
    }
  }
};

//linear time complexity O(n)
binarySearchTreeMethods.contains = function(input){
  var result = false;

  var helper = function(input) {
    if(input === this.value){
      result = true;
    } else if(input < this.value){
      helper.call(this.left, input);
    } else if(input > this.value){
      helper.call(this.right, input);
    }
  };
  helper.call(this, input);
  return result;
};


//linear time complexity O(n)
binarySearchTreeMethods.DFSEach = function(input){
  if(this.value){
    input(this.value);
  }
  if(this.left){
    this.left.DFSEach(input);
  }

  if(this.right){
    this.right.DFSEach(input);
  }
};

//linear time complexity O(n)
binarySearchTreeMethods.DFSToArray = function(input, vals = []){
  if(this.value){
    vals.push(this.value);
  }
  if(this.left){
    this.left.DFSToArray(input, vals);
  }

  if(this.right){
    this.right.DFSToArray(input, vals);
  }
  return vals;
};


binarySearchTreeMethods.BFSEach = function(cb) {
  var current = [this];

  while (current.length > 0) {

    var next = [];

    for (var node of current) {

      cb(node);

      if (node.left) {
        next.push(node.left);
      }
      if (node.right) {
        next.push(node.right);
      }

    current = next;
    }
  }
}

binarySearchTreeMethods.BFSToArray = function(vals = []){
  let current = [this];

  while(current.length > 0){

    let next = [];

    for(var node of current) {
      vals.push(node.value);

      if(node.left){
        next.push(node.left);
      }

      if(node.right){
        next.push(node.right);
      }

    }

    current = next;
  }

  return vals;
};


let valArray = [];
let listArray = [];
var squareIt = function(value) { valArray.push(value*value); };
var listNodes = function(node) { listArray.push(node.value);};

let BST = new BinarySearchTree(10);
BST.insert(9);
BST.insert(6);
BST.insert(5);
BST.insert(8);
BST.contains(5);
BST.insert(15);
BST.insert(13);
BST.insert(20);
BST.contains(20);
BST.DFSToArray();
BST.DFSEach(squareIt);
console.log(valArray)
BST.DFSToArray()
console.log(valArray);
BST.BFSToArray();
BST.BFSEach(listNodes);
console.log(listArray);


module.exports = {
  binarySearchTreeMethods,
  BinarySearchTree
}

File: go/binary_search_tree.go
package main

import (
	"fmt"
)

type bst *BinarySearchTree

type BinarySearchTree struct {
	ValueInfo  string
	Value      int
	ValueComma string
	LeftInfo   string
	Left       *BinarySearchTree
	LeftComma  string
	RightInfo  string
	Right      *BinarySearchTree
}

func InitBinarySearchTree(key int) *BinarySearchTree {
	return &BinarySearchTree{
		ValueInfo:  "Value:",
		Value:      key,
		ValueComma: ",",
		LeftInfo:   "Left:",
		Left:       nil,
		LeftComma:  ",",
		RightInfo:  "Right:",
		Right:      nil,
	}
}

func (b *BinarySearchTree) Insert(key int) {
	if key < b.Value {
		if b.Left == nil {
			b.Left = InitBinarySearchTree(key)
		} else {
			b.Left.Insert(key)
		}
	} else if key > b.Value {
		if b.Right == nil {
			b.Right = InitBinarySearchTree(key)
		} else {
			b.Right.Insert(key)
		}
	}
}

func recurse(result bool, node *BinarySearchTree, input int) bool {
	if node == nil {
		return result
	}

	if input == node.Value {
		result = true
	} else if input < node.Value {
		return recurse(result, node.Left, input)
	} else if input > node.Value {
		return recurse(result, node.Right, input)
	}

	return result
}

func (b *BinarySearchTree) Contains(input int) bool {
	result := false

	if input == b.Value {
		result = true
	} else if input < b.Value && b.Left != nil {
		result = recurse(result, b.Left, input)
	} else if input > b.Value && b.Right != nil {
		result = recurse(result, b.Right, input)
	}

	return result
}

func (b *BinarySearchTree) DepthFirstToArray(v []int) []int {
	if b.Value <= 0 || b.Value > 0 {
		v = append(v, b.Value)
	}

	if b.Left != nil {
		v = b.Left.DepthFirstToArray(v)
	}

	if b.Right != nil {
		v = b.Right.DepthFirstToArray(v)
	}

	return v
}

func (b *BinarySearchTree) BreadthFirstToArray(v []int) []int {
	var current []bst
	var next []bst
	current = make([]bst, 0)
	current = append(current, b)

	for len(current) > 0 {
		next = make([]bst, 0)

		for _, node := range current {

			v = append(v, node.Value)

			if node.Left != nil {
				next = append(next, node.Left)
			}

			if node.Right != nil {
				next = append(next, node.Right)
			}
		}
		current = next
	}

	return v
}

func main() {
	bt := InitBinarySearchTree(100)
	fmt.Println(bt)
	bt.Insert(10)
	fmt.Println(bt)
	bt.Insert(110)
	bt.Insert(20)
  bt.Insert(-10)
  bt.Insert(-11)
  bt.Insert(6)
	bt.Insert(5)
	bt.Insert(135)
	fmt.Println(bt.Contains(20))
	fmt.Println(bt.Contains(19))
	fmt.Println(bt.DepthFirstToArray(make([]int, 0)))
	fmt.Println(bt.BreadthFirstToArray(make([]int, 0)))
}
